File: python/leetcode/286.py
class Solution(object):
def wallsAndGates(self, a):
    for i in range(len(a)):
        for j in range(len(a[0])):
            if a[i][j] == 0:
                stack = [(i + 1, j, 1), (i - 1, j, 1), (i, j + 1, 1), (i, j - 1, 1)]
                while stack:
                    x, y, dist = stack.pop()
                    if x < 0 or y < 0 or x >= len(a) or y >= len(a[0]) or a[x][y] < dist:
                        continue
                    a[x][y] = dist
                    stack.append((x + 1, y, dist + 1))
                    stack.append((x - 1, y, dist + 1))
                    stack.append((x, y + 1, dist + 1))
                    stack.append((x, y - 1, dist + 1))

File: cpp/leetcode/286.cpp
class Solution {
public:
    void wallsAndGates(vector<vector<int>>& rooms) {
        if (rooms.empty()) return;
        int m = rooms.size(), n = rooms[0].size();
        queue<pair<int, int>> q;
        vector<pair<int, int>> dirs = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (!rooms[i][j]) q.push(make_pair(i, j));
            }
        }
        while (!q.empty()) {
            int r = q.front().first, c = q.front().second; q.pop();
            for (auto dir : dirs) {
                int x = r + dir.first, y = c + dir.second;
                if (x < 0 || y < 0 || x >= m || y >= n || rooms[x][y] <= rooms[r][c] + 1) continue;
                rooms[x][y] = rooms[r][c] + 1;
                q.push(make_pair(x, y));
            }
        }
    }
};